checker/src/org/checkerframework/checker/lock/LockAnnotatedTypeFactory.java

```diff
@@ -104,28 +142,353 @@ public LockTransfer createFlowTransferFunction(CFAbstractAnalysis<CFValue, LockS
         return new LockTransfer((LockAnalysis) analysis,(LockChecker)this.checker);
     }
 
-    protected AnnotatedTypeMirror getDeclaredAndDefaultedAnnotatedType(Tree tree) {
-        shouldCache = false;
+    class LockQualifierHierarchy extends MultiGraphQualifierHierarchy {
+
+        public LockQualifierHierarchy(MultiGraphFactory f) {
+            super(f, LOCKHELD);
+        }
+
+        boolean isGuardedBy(AnnotationMirror am) {
+            return AnnotationUtils.areSameIgnoringValues(am, GUARDEDBY);
+        }
+
+        @Override
+        public boolean isSubtype(AnnotationMirror rhs, AnnotationMirror lhs) {
+
+            boolean lhsIsGuardedBy = isGuardedBy(lhs);
+            boolean rhsIsGuardedBy = isGuardedBy(rhs);
+
+            if (lhsIsGuardedBy && rhsIsGuardedBy) {
+                // Two @GuardedBy annotations are considered subtypes of each other if and only if their values match exactly.
+
+                List<String> lhsValues =
+                    AnnotationUtils.getElementValueArray(lhs, "value", String.class, true);
+                List<String> rhsValues =
+                    AnnotationUtils.getElementValueArray(rhs, "value", String.class, true);
+
+                return rhsValues.containsAll(lhsValues) && lhsValues.containsAll(rhsValues);
+            }
+
+            boolean lhsIsGuardSatisfied = AnnotationUtils.areSameIgnoringValues(lhs, GUARDSATISFIED);
+            boolean rhsIsGuardSatisfied = AnnotationUtils.areSameIgnoringValues(rhs, GUARDSATISFIED);
+
+            if (lhsIsGuardSatisfied && rhsIsGuardSatisfied) {
+                // Two @GuardSatisfied annotations are considered subtypes of each other if and only if their indices match exactly.
```

Comment:
I found these comments confusing because two `@GuardStatisfied` annotations are always subtypes according the code in this method. But I see now you are pointing that there are additional constraints on when two `@GS` can assigned.  I would rewrite these comments to indicate that.